sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,309 @@
+input length.
+
+## Bounding Function Output Length
+
+Importantly, our smart contracting language does not allow the
+creation of variable length lists: there are no `list` or
+`cons` constructors, and buffer lengths must be statically
+defined. Under such requirements (and given that recursion is
+illegal), determining the output lengths of functions is rather
+directly achievable. To see this, we'll examine trying to compute the
+output lengths for the only functions allowed to iterate in the
+language:
+
+```
+outputLen(map f list<t>)     := Len(list<t>) * outputLen(f t)
+outputLen(filter f list<t>)  := Len(list<t>)
+outputLen(fold f list<t> s)  := Len(s)
+```
+
+Many functions within the language will output values larger than the
+function's input, _however_, these outputs will be bound by
+statically inferable constants. For example, the data function
+_fetch-entry_ will always return an object whose size is equal
+to the specified value type of the map.
+
+A complete proof for the static runtime analysis of smart contracts
+will be included with the implementation of the language.
+
+# Deploying the Smart Contract
+
+Smart contracts on the Stacks blockchain will be deployed directly as
+source code. The goal of the smart contracting language is that the
+code of the contract defines the _ground truth_ about the intended
+functionality of the contract. While seemingly banal, many systems
+chose instead to use a a compiler to translate from a a friendly
+high-level language to a lower-level language deployed on the
+blockchain. Such an architecture is needlessly dangerous. A bug in
+such a compiler could lead to a bug in a deployed smart contract when
+no such bug exists in the original source. This is problematic for
+recovery --- a hard fork to "undo" any should-have-been invalid
+transactions would be contentious and potentially create a rift in the
+community, especially as it will not be easy to deduce which contracts
+exactly were affected and for how long. In contrast, bugs in the VM
+itself present a more clear case for a hard fork: the smart contract
+was defined correctly, as everyone can see directly on the chain, but
+illegal transactions were incorrectly marked as valid.
+
+# Example: Simple Naming System
+
+To demonstrate the expressiveness of this smart contracting language,
+let's look at an example smart contract which implements a simple
+naming system with just two kinds of transactions: _preorder_
+and _register_. The requirements of the system are as follows:
+
+1. Names may only be owned by one principal
+2. A register is only allowed if there is a corresponding preorder
+   with a matching hash
+3. A register transaction must be signed by the same principal who
+   paid for the preorder
+4. A preorder must have paid at least the price of the name. Names
+   are represented as integers, and any name less than 100000 costs
+   1000 microstacks, while all other names cost 100 microstacks.
```

Comment:
5. Preorder hashes must be globally unique.

Reply:
Added to the list, and updated the smart contract.